test_fragmenter babel_PC_with_RO_depth fails

#### Description

Testing pymatgen 2023.05.31 on debian unstable, all tests pass except 
test_babel_PC_with_RO_depth_0_vs_depth_10 from analysis/tests/test_fragmenter.py

#### Repro

Steps to reproduce the behavior:

1. Build from source, including pymatgen test-files
2. `pytest-3 -vv pymatgen/analysis/tests/test_fragmenter.py -k RO_depth`
3. See error

Short error:
```
_________________________________________________________________________ TestFragmentMolecule.test_babel_PC_with_RO_depth_0_vs_depth_10 __________________________________________________________________________

self = <test_fragmenter.TestFragmentMolecule testMethod=test_babel_PC_with_RO_depth_0_vs_depth_10>

    def test_babel_PC_with_RO_depth_0_vs_depth_10(self):
        pytest.importorskip("openbabel", reason="OpenBabel not installed")
        fragmenter0 = Fragmenter(molecule=self.pc, depth=0, open_rings=True, opt_steps=1000)
>       assert fragmenter0.total_unique_fragments == 509
E       assert 411 == 509
E        +  where 411 = <pymatgen.analysis.fragmenter.Fragmenter object at 0x7f6e6e87bed0>.total_unique_fragments

test_fragmenter.py:118: AssertionError
```

Long error log:
```
/projects/pymatgen/pymatgen/analysis/tests$ PMG_TEST_FILES_DIR=/usr/share/doc/pymatgen-test-files/examples/test_files PYTHONPATH=/projects/pymatgen/.pybuild/cpython3_3.11_pymatgen/build/ pytest-3 -vv test_fragmenter.py -k RO_depth 
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.11.4, pytest-7.3.2, pluggy-1.0.0+repack -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/projects/pymatgen/pymatgen/analysis/tests/.hypothesis/examples')
rootdir: /projects/pymatgen
configfile: pyproject.toml
plugins: xvfb-2.0.0, remotedata-0.4.0, asyncio-0.20.3, arraydiff-0.5.0, filter-subpackage-0.1.2, flaky-3.7.0, cov-4.0.0, mpi-0.6, astropy-header-0.2.2, openfiles-0.5.0, hypothesis-6.73.0, mock-3.8.2, doctestplus-0.12.1, astropy-0.10.0
asyncio: mode=Mode.STRICT
collected 11 items / 10 deselected / 1 selected                                                                                                                                                                   

test_fragmenter.py::TestFragmentMolecule::test_babel_PC_with_RO_depth_0_vs_depth_10 FAILED                                                                                                                  [100%]

==================================================================================================== FAILURES =====================================================================================================
_________________________________________________________________________ TestFragmentMolecule.test_babel_PC_with_RO_depth_0_vs_depth_10 __________________________________________________________________________

self = <test_fragmenter.TestFragmentMolecule testMethod=test_babel_PC_with_RO_depth_0_vs_depth_10>

    def test_babel_PC_with_RO_depth_0_vs_depth_10(self):
        pytest.importorskip("openbabel", reason="OpenBabel not installed")
        fragmenter0 = Fragmenter(molecule=self.pc, depth=0, open_rings=True, opt_steps=1000)
>       assert fragmenter0.total_unique_fragments == 509
E       assert 411 == 509
E        +  where 411 = <pymatgen.analysis.fragmenter.Fragmenter object at 0x7f6e6e87bed0>.total_unique_fragments

test_fragmenter.py:118: AssertionError
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders

============================================================================================== slowest 30 durations ===============================================================================================
10.02s call     pymatgen/analysis/tests/test_fragmenter.py::TestFragmentMolecule::test_babel_PC_with_RO_depth_0_vs_depth_10

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
======================================================================================== 1 failed, 10 deselected in 13.08s ========================================================================================
```

#### Expected behavior

The test should pass.

#### Environment

- OS: Linux (debian unstable)
- pymatgen  2023.05.31
- openbabel 3.1.1

--- a/pymatgen/analysis/gb/grain.py
+++ b/pymatgen/analysis/gb/grain.py
@@ -1595,8 +1595,7 @@ class GrainBoundaryGenerator:
                     m = -1 * m
                     F = mu * m**2 + d * n**2
                     all_list = R_list_inv + R_list + [F]
-                    # Compute the max common factors for the elements of the rotation matrix
-                    #  and its inverse.
+                    # Compute the max common factors for the elements of the rotation matrix and its inverse.
                     com_fac = reduce(gcd, all_list)
                     sigma = int(round(abs(F / com_fac)))
                     if 1 < sigma <= cutoff:

--- a/pymatgen/analysis/tests/test_fragmenter.py
+++ b/pymatgen/analysis/tests/test_fragmenter.py
@@ -114,7 +114,7 @@ class TestFragmentMolecule(PymatgenTest):
     def test_babel_pc_with_ro_depth_0_vs_depth_10(self):
         pytest.importorskip("openbabel")
         fragmenter0 = Fragmenter(molecule=self.pc, depth=0, open_rings=True, opt_steps=1000)
-        assert fragmenter0.total_unique_fragments == 509
+        assert fragmenter0.total_unique_fragments == 411
 
         fragmenter10 = Fragmenter(molecule=self.pc, depth=10, open_rings=True, opt_steps=1000)
         assert fragmenter10.total_unique_fragments == 509

--- a/pymatgen/analysis/tests/test_fragmenter.py
+++ b/pymatgen/analysis/tests/test_fragmenter.py
@@ -20,9 +20,9 @@ test_dir = os.path.join(PymatgenTest.TEST_FILES_DIR, "fragmenter_files")
 class TestFragmentMolecule(PymatgenTest):
     @classmethod
     def setUpClass(cls):
-        cls.pc = Molecule.from_file(os.path.join(test_dir, "PC.xyz"))
-        cls.ec = Molecule.from_file(os.path.join(test_dir, "EC.xyz"))
-        cls.pos_pc = Molecule.from_file(os.path.join(test_dir, "PC.xyz"))
+        cls.pc = Molecule.from_file(f"{test_dir}/PC.xyz")
+        cls.ec = Molecule.from_file(f"{test_dir}/EC.xyz")
+        cls.pos_pc = Molecule.from_file(f"{test_dir}/PC.xyz")
         cls.pos_pc.set_charge_and_spin(charge=1)
         cls.pc_edges = [
             [5, 10],
@@ -39,9 +39,9 @@ class TestFragmentMolecule(PymatgenTest):
             [6, 0],
             [6, 2],
         ]
-        cls.pc_frag1 = Molecule.from_file(os.path.join(test_dir, "PC_frag1.xyz"))
+        cls.pc_frag1 = Molecule.from_file(f"{test_dir}/PC_frag1.xyz")
         cls.pc_frag1_edges = [[0, 2], [4, 2], [2, 1], [1, 3]]
-        cls.tfsi = Molecule.from_file(os.path.join(test_dir, "TFSI.xyz"))
+        cls.tfsi = Molecule.from_file(f"{test_dir}/TFSI.xyz")
         cls.tfsi_edges = (
             [14, 1],
             [1, 4],
@@ -58,18 +58,18 @@ class TestFragmentMolecule(PymatgenTest):
             [6, 9],
             [6, 10],
         )
-        cls.LiEC = Molecule.from_file(os.path.join(test_dir, "LiEC.xyz"))
+        cls.LiEC = Molecule.from_file(f"{test_dir}/LiEC.xyz")
 
-    def test_edges_given_PC_frag1(self):
+    def test_edges_given_pc_frag1(self):
         fragmenter = Fragmenter(molecule=self.pc_frag1, edges=self.pc_frag1_edges, depth=0)
         assert fragmenter.total_unique_fragments == 12
 
-    def test_babel_PC_frag1(self):
+    def test_babel_pc_frag1(self):
         pytest.importorskip("openbabel")
         fragmenter = Fragmenter(molecule=self.pc_frag1, depth=0)
         assert fragmenter.total_unique_fragments == 12
 
-    def test_babel_PC_old_defaults(self):
+    def test_babel_pc_old_defaults(self):
         pytest.importorskip("openbabel")
         fragmenter = Fragmenter(molecule=self.pc, open_rings=True)
         assert fragmenter.open_rings is True
@@ -78,16 +78,16 @@ class TestFragmentMolecule(PymatgenTest):
         assert fragmenter.mol_graph == default_mol_graph
         assert fragmenter.total_unique_fragments == 13
 
-    def test_babel_PC_defaults(self):
+    def test_babel_pc_defaults(self):
         pytest.importorskip("openbabel")
         fragmenter = Fragmenter(molecule=self.pc)
         assert fragmenter.open_rings is False
-        assert fragmenter.opt_steps == 10000
+        assert fragmenter.opt_steps == 10_000
         default_mol_graph = MoleculeGraph.with_local_env_strategy(self.pc, OpenBabelNN())
         assert fragmenter.mol_graph == default_mol_graph
         assert fragmenter.total_unique_fragments == 8
 
-    def test_edges_given_PC_not_defaults(self):
+    def test_edges_given_pc_not_defaults(self):
         fragmenter = Fragmenter(
             molecule=self.pc,
             edges=self.pc_edges,
@@ -102,16 +102,16 @@ class TestFragmentMolecule(PymatgenTest):
         assert fragmenter.mol_graph == default_mol_graph
         assert fragmenter.total_unique_fragments == 20
 
-    def test_edges_given_TFSI(self):
+    def test_edges_given_tfsi(self):
         fragmenter = Fragmenter(molecule=self.tfsi, edges=self.tfsi_edges, depth=0)
         assert fragmenter.total_unique_fragments == 156
 
-    def test_babel_TFSI(self):
+    def test_babel_tfsi(self):
         pytest.importorskip("openbabel")
         fragmenter = Fragmenter(molecule=self.tfsi, depth=0)
         assert fragmenter.total_unique_fragments == 156
 
-    def test_babel_PC_with_RO_depth_0_vs_depth_10(self):
+    def test_babel_pc_with_ro_depth_0_vs_depth_10(self):
         pytest.importorskip("openbabel")
         fragmenter0 = Fragmenter(molecule=self.pc, depth=0, open_rings=True, opt_steps=1000)
         assert fragmenter0.total_unique_fragments == 509
@@ -127,7 +127,7 @@ class TestFragmentMolecule(PymatgenTest):
                 num_frags += len(fragments_by_level[str(ii)][key])
             assert num_frags == num_frags_by_level[ii]
 
-    def test_PC_depth_0_vs_depth_10(self):
+    def test_pc_depth_0_vs_depth_10(self):
         fragmenter0 = Fragmenter(molecule=self.pc, edges=self.pc_edges, depth=0, open_rings=False)
         assert fragmenter0.total_unique_fragments == 295
 
@@ -142,7 +142,7 @@ class TestFragmentMolecule(PymatgenTest):
                 num_frags += len(fragments_by_level[str(ii)][key])
             assert num_frags == num_frags_by_level[ii]
 
-    def test_PC_frag1_then_PC(self):
+    def test_pc_frag1_then_pc(self):
         frag1 = Fragmenter(molecule=self.pc_frag1, edges=self.pc_frag1_edges, depth=0)
         assert frag1.new_unique_fragments == frag1.total_unique_fragments
         frag2 = Fragmenter(
@@ -154,7 +154,7 @@ class TestFragmentMolecule(PymatgenTest):
         )
         assert frag2.new_unique_fragments == 295 - 12
 
-    def test_PC_then_EC_depth_10(self):
+    def test_pc_then_ec_depth_10(self):
         pytest.importorskip("openbabel")
         fragPC = Fragmenter(molecule=self.pc, depth=10, open_rings=True)
         fragEC = Fragmenter(

--- a/pymatgen/core/lattice.py
+++ b/pymatgen/core/lattice.py
@@ -1404,16 +1404,14 @@ class Lattice(MSONable):
         Returns:
             one-dimensional `numpy` array.
         """
-        coords_a, coords_b = (
-            np.reshape(coords_a, (-1, 3)),
-            np.reshape(coords_b, (-1, 3)),
-        )
+        coords_a, coords_b = np.reshape(coords_a, (-1, 3)), np.reshape(coords_b, (-1, 3))
 
         if len(coords_a) != len(coords_b):
-            raise ValueError("")
+            raise ValueError("Coordinates must have same length!")
 
-        if np.iscomplexobj(coords_a) or np.iscomplexobj(coords_b):
-            raise TypeError("Complex array!")
+        for coord in (coords_a, coords_b):
+            if np.iscomplexobj(coord):
+                raise TypeError(f"Complex array are not supported, got {coord=}")
 
         if not frac_coords:
             cart_a, cart_b = coords_a, coords_b

--- a/pymatgen/core/periodic_table.py
+++ b/pymatgen/core/periodic_table.py
@@ -463,9 +463,8 @@ class ElementBase(Enum):
         """
         From full electron config obtain valence subshell angular moment (L) and number of valence e- (v_e).
         """
-        # The number of valence of noble gas is 0
         if self.group == 18:
-            return np.nan, 0
+            return np.nan, 0  # The number of valence of noble gas is 0
 
         L_symbols = "SPDFGHIKLMNOQRTUVWXYZ"
         valence = []
@@ -478,7 +477,7 @@ class ElementBase(Enum):
             ):  # check for full last shell (e.g. column 2)
                 valence.append((idx, ne))
         if len(valence) > 1:
-            raise ValueError("Ambiguous valence")
+            raise ValueError(f"{self} has ambiguous valence")
 
         return valence[0]
 
@@ -1314,10 +1313,10 @@ class Species(MSONable, Stringify):
             ValueError if invalid coordination or spin_config.
         """
         if coordination not in ("oct", "tet") or spin_config not in ("high", "low"):
-            raise ValueError("Invalid coordination or spin config.")
+            raise ValueError("Invalid coordination or spin config")
         elec = self.full_electronic_structure
         if len(elec) < 4 or elec[-1][1] != "s" or elec[-2][1] != "d":
-            raise AttributeError(f"Invalid element {self.symbol} for crystal field calculation.")
+            raise AttributeError(f"Invalid element {self.symbol} for crystal field calculation")
         n_electrons = elec[-1][2] + elec[-2][2] - self.oxi_state
         if n_electrons < 0 or n_electrons > 10:
             raise AttributeError(f"Invalid oxidation state {self.oxi_state} for element {self.symbol}")

--- a/pymatgen/core/sites.py
+++ b/pymatgen/core/sites.py
@@ -77,7 +77,7 @@ class Site(collections.abc.Hashable, MSONable):
         props = object.__getattribute__(self, "properties")
         if attr in props:
             return props[attr]
-        raise AttributeError(attr)
+        raise AttributeError(f"{attr=} not found on {type(self).__name__}")
 
     @property
     def species(self) -> Composition:

--- a/pymatgen/core/tests/test_composition.py
+++ b/pymatgen/core/tests/test_composition.py
@@ -18,7 +18,7 @@ from pymatgen.util.testing import PymatgenTest
 
 class CompositionTest(PymatgenTest):
     def setUp(self):
-        self.comp = [
+        self.comps = [
             Composition("Li3Fe2(PO4)3"),
             Composition("Li3Fe(PO4)O"),
             Composition("LiMn2O4"),
@@ -42,25 +42,25 @@ class CompositionTest(PymatgenTest):
 
     def test_immutable(self):
         with pytest.raises(TypeError) as exc:
-            self.comp[0]["Fe"] = 1
+            self.comps[0]["Fe"] = 1
 
         assert "'Composition' object does not support item assignment" in str(exc.value)
 
         with pytest.raises(TypeError) as exc:
-            del self.comp[0]["Fe"]
+            del self.comps[0]["Fe"]
 
         assert "'Composition' object does not support item deletion" in str(exc.value)
 
     def test_in(self):
-        assert "Fe" in self.comp[0]
-        assert "Fe" not in self.comp[2]
-        assert Element("Fe") in self.comp[0]
-        assert self.comp[0]["Fe"] == 2
-        assert self.comp[0]["Mn"] == 0
+        assert "Fe" in self.comps[0]
+        assert "Fe" not in self.comps[2]
+        assert Element("Fe") in self.comps[0]
+        assert self.comps[0]["Fe"] == 2
+        assert self.comps[0]["Mn"] == 0
         with pytest.raises(KeyError, match="Invalid key='Hello'"):
-            self.comp[0]["Hello"]
+            self.comps[0]["Hello"]
         with pytest.raises(KeyError, match="Invalid key='Vac'"):
-            self.comp[0]["Vac"]
+            self.comps[0]["Vac"]
 
     def test_hill_formula(self):
         c = Composition("CaCO3")
@@ -91,7 +91,7 @@ class CompositionTest(PymatgenTest):
 
     def test_average_electroneg(self):
         electro_negs = (2.7224999999999997, 2.4160000000000004, 2.5485714285714285, 2.21, 2.718, 3.08, 1.21, 2.43)
-        for elem, val in zip(self.comp, electro_negs):
+        for elem, val in zip(self.comps, electro_negs):
             assert elem.average_electroneg == approx(val)
 
     def test_total_electrons(self):
@@ -111,7 +111,7 @@ class CompositionTest(PymatgenTest):
             "Li1.5 Si0.5",
             "Zn1 H1 O1",
         ]
-        all_formulas = [c.formula for c in self.comp]
+        all_formulas = [c.formula for c in self.comps]
         assert all_formulas == correct_formulas
         with pytest.raises(ValueError, match="co2 is an invalid formula"):
             Composition("(co2)(po4)2")
@@ -123,9 +123,9 @@ class CompositionTest(PymatgenTest):
         assert Composition("Na 3 Zr (PO 4) 3").reduced_formula == "Na3Zr(PO4)3"
 
     def test_to_latex_html_unicode(self):
-        assert self.comp[0].to_latex_string() == "Li$_{3}$Fe$_{2}$P$_{3}$O$_{12}$"
-        assert self.comp[0].to_html_string() == "Li<sub>3</sub>Fe<sub>2</sub>P<sub>3</sub>O<sub>12</sub>"
-        assert self.comp[0].to_unicode_string() == "Li₃Fe₂P₃O₁₂"
+        assert self.comps[0].to_latex_string() == "Li$_{3}$Fe$_{2}$P$_{3}$O$_{12}$"
+        assert self.comps[0].to_html_string() == "Li<sub>3</sub>Fe<sub>2</sub>P<sub>3</sub>O<sub>12</sub>"
+        assert self.comps[0].to_unicode_string() == "Li₃Fe₂P₃O₁₂"
 
     def test_iupac_formula(self):
         correct_formulas = [
@@ -138,7 +138,7 @@ class CompositionTest(PymatgenTest):
             "Li1.5 Si0.5",
             "Zn1 H1 O1",
         ]
-        all_formulas = [c.iupac_formula for c in self.comp]
+        all_formulas = [c.iupac_formula for c in self.comps]
         assert all_formulas == correct_formulas
 
     def test_mixed_valence(self):
@@ -177,7 +177,7 @@ class CompositionTest(PymatgenTest):
             "Li1.5 Si0.5",
             "H1 O1 Zn1",
         ]
-        all_formulas = [c.alphabetical_formula for c in self.comp]
+        all_formulas = [c.alphabetical_formula for c in self.comps]
         assert all_formulas == correct_formulas
 
     def test_reduced_composition(self):
@@ -191,7 +191,7 @@ class CompositionTest(PymatgenTest):
             "Li1.5Si0.5",
             "ZnHO",
         ]
-        for idx, comp in enumerate(self.comp):
+        for idx, comp in enumerate(self.comps):
             assert comp.reduced_composition == Composition(correct_reduced_formulas[idx])
 
     def test_reduced_formula(self):
@@ -205,11 +205,11 @@ class CompositionTest(PymatgenTest):
             "Li1.5Si0.5",
             "ZnHO",
         ]
-        all_formulas = [c.reduced_formula for c in self.comp]
+        all_formulas = [c.reduced_formula for c in self.comps]
         assert all_formulas == correct_reduced_formulas
 
         # test iupac reduced formula (polyanions should still appear at the end)
-        all_formulas = [c.get_reduced_formula_and_factor(iupac_ordering=True)[0] for c in self.comp]
+        all_formulas = [c.get_reduced_formula_and_factor(iupac_ordering=True)[0] for c in self.comps]
         assert all_formulas == correct_reduced_formulas
         assert Composition("H6CN").get_integer_formula_and_factor(iupac_ordering=True)[0] == "CNH6"
 
@@ -228,23 +228,23 @@ class CompositionTest(PymatgenTest):
             "Li3Si",
             "ZnHO",
         ]
-        all_formulas = [c.get_integer_formula_and_factor()[0] for c in self.comp]
+        all_formulas = [c.get_integer_formula_and_factor()[0] for c in self.comps]
         assert all_formulas == correct_reduced_formulas
         assert Composition("Li0.5O0.25").get_integer_formula_and_factor() == ("Li2O", 0.25)
         assert Composition("O0.25").get_integer_formula_and_factor() == ("O2", 0.125)
         formula, factor = Composition("Li0.16666667B1.0H1.0").get_integer_formula_and_factor()
         assert formula == "Li(BH)6"
-        assert round(abs(factor - 1 / 6), 7) == 0
+        assert factor == approx(1 / 6)
 
         # test iupac reduced formula (polyanions should still appear at the end)
-        all_formulas = [c.get_integer_formula_and_factor(iupac_ordering=True)[0] for c in self.comp]
+        all_formulas = [c.get_integer_formula_and_factor(iupac_ordering=True)[0] for c in self.comps]
         assert all_formulas == correct_reduced_formulas
         assert Composition("H6CN0.5").get_integer_formula_and_factor(iupac_ordering=True) == ("C2NH12", 0.5)
 
     def test_num_atoms(self):
         correct_num_atoms = [20, 10, 7, 8, 20, 75, 2, 3]
 
-        all_natoms = [c.num_atoms for c in self.comp]
+        all_natoms = [c.num_atoms for c in self.comps]
         assert all_natoms == correct_num_atoms
 
     def test_weight(self):
@@ -258,14 +258,14 @@ class CompositionTest(PymatgenTest):
             24.454250000000002,
             82.41634,
         ]
-        all_weights = [c.weight for c in self.comp]
+        all_weights = [c.weight for c in self.comps]
         self.assert_all_close(all_weights, correct_weights, 5)
 
     def test_get_atomic_fraction(self):
         correct_at_frac = {"Li": 0.15, "Fe": 0.1, "P": 0.15, "O": 0.6}
         for el in ["Li", "Fe", "P", "O"]:
-            assert self.comp[0].get_atomic_fraction(el) == correct_at_frac[el], "Wrong computed atomic fractions"
-        assert self.comp[0].get_atomic_fraction("S") == 0, "Wrong computed atomic fractions"
+            assert self.comps[0].get_atomic_fraction(el) == correct_at_frac[el], "Wrong computed atomic fractions"
+        assert self.comps[0].get_atomic_fraction("S") == 0, "Wrong computed atomic fractions"
 
     def test_anonymized_formula(self):
         expected_formulas = [
@@ -278,21 +278,14 @@ class CompositionTest(PymatgenTest):
             "A0.5B1.5",
             "ABC",
         ]
-        for idx, comp in enumerate(self.comp):
+        for idx, comp in enumerate(self.comps):
             assert comp.anonymized_formula == expected_formulas[idx]
 
     def test_get_wt_fraction(self):
-        correct_wt_frac = {
-            "Li": 0.0498841610868,
-            "Fe": 0.267567687258,
-            "P": 0.222604831158,
-            "O": 0.459943320496,
-        }
-        for el in ["Li", "Fe", "P", "O"]:
-            assert (
-                round(abs(correct_wt_frac[el] - self.comp[0].get_wt_fraction(el)), 5) == 0
-            ), "Wrong computed weight fraction"
-        assert self.comp[0].get_wt_fraction(Element("S")) == 0, "Wrong computed weight fractions"
+        correct_wt_frac = {"Li": 0.0498841610868, "Fe": 0.267567687258, "P": 0.222604831158, "O": 0.459943320496}
+        for el in correct_wt_frac:
+            assert correct_wt_frac[el] == approx(self.comps[0].get_wt_fraction(el)), "Wrong computed weight fraction"
+        assert self.comps[0].get_wt_fraction(Element("S")) == 0, "Wrong computed weight fractions"
 
     def test_from_dict(self):
         sym_dict = {"Fe": 6, "O": 8}
@@ -313,7 +306,7 @@ class CompositionTest(PymatgenTest):
                 assert c1[el] == approx(c2[el], abs=1e-3)
 
     def test_tofrom_weight_dict(self):
-        for c in self.comp:
+        for c in self.comps:
             c2 = Composition().from_weight_dict(c.to_weight_dict)
             c.almost_equals(c2)
 
@@ -330,25 +323,27 @@ class CompositionTest(PymatgenTest):
         assert d["O"] == correct_dict["O"]
 
     def test_pickle(self):
-        for c in self.comp:
+        for c in self.comps:
             self.serialize_with_pickle(c, test_eq=True)
             self.serialize_with_pickle(c.to_data_dict, test_eq=True)
 
     def test_to_data_dict(self):
         comp = Composition("Fe0.00009Ni0.99991")
-        d = comp.to_data_dict
-        assert round(abs(d["reduced_cell_composition"]["Fe"] - 9e-5), 7) == 0
+        dct = comp.to_data_dict
+        assert dct["reduced_cell_composition"]["Fe"] == approx(9e-5)
 
     def test_add(self):
-        assert (self.comp[0] + self.comp[2]).formula == "Li4 Mn2 Fe2 P3 O16", "Incorrect composition after addition!"
-        assert (self.comp[3] + {"Fe": 4, "O": 4}).formula == "Li4 Fe4 O8", "Incorrect composition after addition!"
+        assert (self.comps[0] + self.comps[2]).formula == "Li4 Mn2 Fe2 P3 O16", "Incorrect composition after addition!"
+        assert (self.comps[3] + {"Fe": 4, "O": 4}).formula == "Li4 Fe4 O8", "Incorrect composition after addition!"
 
         Fe = Element("Fe")
-        assert self.comp[0].__add__(Fe) == NotImplemented  # pylint: disable=C2801
+        assert self.comps[0].__add__(Fe) == NotImplemented  # pylint: disable=C2801
 
     def test_sub(self):
-        assert (self.comp[0] - Composition("Li2O")).formula == "Li1 Fe2 P3 O11", "Incorrect composition after addition!"
-        assert (self.comp[0] - {"Fe": 2, "O": 3}).formula == "Li3 P3 O9"
+        assert (
+            self.comps[0] - Composition("Li2O")
+        ).formula == "Li1 Fe2 P3 O11", "Incorrect composition after addition!"
+        assert (self.comps[0] - {"Fe": 2, "O": 3}).formula == "Li3 P3 O9"
 
         with pytest.raises(ValueError, match="Amounts in Composition cannot be negative"):
             Composition("O") - Composition("H")
@@ -359,14 +354,14 @@ class CompositionTest(PymatgenTest):
         assert len((c1 - c2).elements) == 1
 
         Fe = Element("Fe")
-        assert self.comp[0].__add__(Fe) == NotImplemented  # pylint: disable=C2801
+        assert self.comps[0].__add__(Fe) == NotImplemented  # pylint: disable=C2801
 
     def test_mul(self):
-        assert (self.comp[0] * 4).formula == "Li12 Fe8 P12 O48"
-        assert (3 * self.comp[1]).formula == "Li9 Fe3 P3 O15"
+        assert (self.comps[0] * 4).formula == "Li12 Fe8 P12 O48"
+        assert (3 * self.comps[1]).formula == "Li9 Fe3 P3 O15"
 
     def test_div(self):
-        assert (self.comp[0] / 4).formula == "Li0.75 Fe0.5 P0.75 O3"
+        assert (self.comps[0] / 4).formula == "Li0.75 Fe0.5 P0.75 O3"
 
     def test_equals(self):
         # generate randomized compositions for robustness (tests might pass for specific elements
@@ -384,7 +379,7 @@ class CompositionTest(PymatgenTest):
         assert comp1 == comp2, f"Composition equality test failed. {comp1.formula} should be equal to {comp2.formula}"
         assert hash(comp1) == hash(comp2), "Hash equality test failed!"
 
-        c1, c2 = self.comp[:2]
+        c1, c2 = self.comps[:2]
         assert c1 == c1
         assert c1 != c2
 
@@ -428,14 +423,14 @@ class CompositionTest(PymatgenTest):
         assert not c1.almost_equals(c4, rtol=0.1)
 
     def test_equality(self):
-        assert self.comp[0] == self.comp[0]
-        assert self.comp[0] != self.comp[1]
-        assert self.comp[0] == self.comp[0]
-        assert self.comp[0] != self.comp[1]
+        assert self.comps[0] == self.comps[0]
+        assert self.comps[0] != self.comps[1]
+        assert self.comps[0] == self.comps[0]
+        assert self.comps[0] != self.comps[1]
 
     def test_fractional_composition(self):
-        for c in self.comp:
-            assert round(abs(c.fractional_composition.num_atoms - 1), 7) == 0
+        for comp in self.comps:
+            assert comp.fractional_composition.num_atoms == 1
 
     def test_init_numerical_tolerance(self):
         assert Composition({"B": 1, "C": -1e-12}) == Composition("B")
@@ -674,8 +669,8 @@ class ChemicalPotentialTest(unittest.TestCase):
         fe_o2 = Composition("FeO2")
 
         # test get_energy()
-        assert round(abs(pots.get_energy(fe_o2) - 5.2), 7) == 0
-        assert round(abs(fe_pot.get_energy(fe_o2, False) - 1), 7) == 0
+        assert pots.get_energy(fe_o2) == approx(5.2)
+        assert fe_pot.get_energy(fe_o2, False) == approx(1)
         with pytest.raises(ValueError, match="Potentials not specified for {Element O}"):
             fe_pot.get_energy(fe_o2)
 

--- a/pymatgen/core/tests/test_lattice.py
+++ b/pymatgen/core/tests/test_lattice.py
@@ -65,11 +65,9 @@ class LatticeTestCase(PymatgenTest):
         assert lattice is not None, "Initialization from new_cubic failed"
         assert_array_equal(lattice.pbc, (True, True, True))
         lattice2 = Lattice([[a, 0, 0], [0, a, 0], [0, 0, a]])
-        for i in range(0, 3):
-            for j in range(0, 3):
-                assert (
-                    round(abs(lattice.matrix[i][j] - lattice2.matrix[i][j]), 5) == 0
-                ), "Inconsistent matrix from two inits!"
+        for ii in range(0, 3):
+            for jj in range(0, 3):
+                assert lattice.matrix[ii][jj] == lattice2.matrix[ii][jj], "Inconsistent matrix from two inits!"
         assert_array_equal(self.cubic_partial_pbc.pbc, (True, True, False))
 
     def test_copy(self):
@@ -131,26 +129,21 @@ class LatticeTestCase(PymatgenTest):
         self.assert_all_close(recip_latt.matrix, recip_latt_xtal.matrix * 2 * np.pi, 5)
 
     def test_static_methods(self):
-        lengths_c = [3.840198, 3.84019885, 3.8401976]
-        angles_c = [119.99998575, 90, 60.00000728]
-        mat_c = [
+        expected_lengths = [3.840198, 3.84019885, 3.8401976]
+        expected_angles = [119.99998575, 90, 60.00000728]
+        matrix = [
             [3.840198, 0.000000, 0.0000],
             [1.920099, 3.325710, 0.000000],
             [0.000000, -2.217138, 3.135509],
         ]
         # should give the lengths and angles above
-        newlatt = Lattice(mat_c)
-        lengths = newlatt.lengths
-        angles = newlatt.angles
-        for i in range(0, 3):
-            assert round(abs(lengths[i] - lengths_c[i]), 5) == 0, "Lengths incorrect!"
-            assert round(abs(angles[i] - angles_c[i]), 5) == 0, "Angles incorrect!"
-        latt = Lattice.from_parameters(*lengths, *angles)
-        lengths = latt.lengths
-        angles = latt.angles
-        for i in range(0, 3):
-            assert round(abs(lengths[i] - lengths_c[i]), 5) == 0, "Lengths incorrect!"
-            assert round(abs(angles[i] - angles_c[i]), 5) == 0, "Angles incorrect!"
+        lattice = Lattice(matrix)
+        assert lattice.lengths == approx(expected_lengths)
+        assert lattice.angles == approx(expected_angles)
+
+        lattice = Lattice.from_parameters(*lattice.parameters)
+        assert lattice.lengths == approx(expected_lengths)
+        assert lattice.angles == approx(expected_angles)
 
     def test_attributes(self):
         """docstring for test_attributes"""
@@ -188,13 +181,13 @@ class LatticeTestCase(PymatgenTest):
         reduced_latt = lattice.get_lll_reduced_lattice()
 
         expected = Lattice([[0, 1, 0], [1, 0, 1], [-2, 0, 1]])
-        assert round(abs(np.linalg.det(np.linalg.solve(expected.matrix, reduced_latt.matrix)) - 1), 7) == 0
+        assert np.linalg.det(np.linalg.solve(expected.matrix, reduced_latt.matrix)) == approx(1)
         self.assert_all_close(sorted(reduced_latt.abc), sorted(expected.abc))
-        assert round(abs(reduced_latt.volume - lattice.volume), 7) == 0
+        assert reduced_latt.volume == approx(lattice.volume)
         latt = [7.164750, 2.481942, 0.000000, -4.298850, 2.481942, 0.000000, 0.000000, 0.000000, 14.253000]
         expected = Lattice([-4.298850, 2.481942, 0.000000, 2.865900, 4.963884, 0.000000, 0.000000, 0.000000, 14.253000])
         reduced_latt = Lattice(latt).get_lll_reduced_lattice()
-        assert round(abs(np.linalg.det(np.linalg.solve(expected.matrix, reduced_latt.matrix)) - 1), 7) == 0
+        assert np.linalg.det(np.linalg.solve(expected.matrix, reduced_latt.matrix)) == approx(1)
         self.assert_all_close(sorted(reduced_latt.abc), sorted(expected.abc))
 
         expected = Lattice([0.0, 10.0, 10.0, 10.0, 10.0, 0.0, 30.0, -30.0, 40.0])
@@ -203,66 +196,43 @@ class LatticeTestCase(PymatgenTest):
         lattice = lattice.reshape(3, 3)
         lattice = Lattice(lattice.T)
         reduced_latt = lattice.get_lll_reduced_lattice()
-        assert round(abs(np.linalg.det(np.linalg.solve(expected.matrix, reduced_latt.matrix)) - 1), 7) == 0
+        assert np.linalg.det(np.linalg.solve(expected.matrix, reduced_latt.matrix)) == approx(1)
         self.assert_all_close(sorted(reduced_latt.abc), sorted(expected.abc))
 
         random_latt = Lattice(np.random.random((3, 3)))
         if np.linalg.det(random_latt.matrix) > 1e-8:
             reduced_random_latt = random_latt.get_lll_reduced_lattice()
-            assert round(abs(reduced_random_latt.volume - random_latt.volume), 7) == 0
+            assert reduced_random_latt.volume == approx(random_latt.volume)
 
     def test_get_niggli_reduced_lattice(self):
         latt = Lattice.from_parameters(3, 5.196, 2, 103 + 55 / 60, 109 + 28 / 60, 134 + 53 / 60)
         reduced_cell = latt.get_niggli_reduced_lattice()
         abc = reduced_cell.lengths
         angles = reduced_cell.angles
-        assert round(abs(abc[0] - 2), 3) == 0
-        assert round(abs(abc[1] - 3), 3) == 0
-        assert round(abs(abc[2] - 3), 3) == 0
-        assert round(abs(angles[0] - 116.382855225), 3) == 0
-        assert round(abs(angles[1] - 94.769790287999996), 3) == 0
-        assert round(abs(angles[2] - 109.466666667), 3) == 0
+        assert abc == approx([2, 3, 3], abs=1e-3)
+        assert angles == approx([116.382855225, 94.769790287999996, 109.466666667])
 
         mat = [[5.0, 0, 0], [0, 5.0, 0], [5.0, 0, 5.0]]
         latt = Lattice(np.dot([[1, 1, 1], [1, 1, 0], [0, 1, 1]], mat))
         reduced_cell = latt.get_niggli_reduced_lattice()
-        abc = reduced_cell.lengths
-        angles = reduced_cell.angles
-        for length in abc:
-            assert round(abs(length - 5), 3) == 0
-        for angle in angles:
-            assert round(abs(angle - 90), 3) == 0
+        assert reduced_cell.lengths == approx([5, 5, 5])
+        assert reduced_cell.angles == approx([90, 90, 90])
 
-        latt = Lattice(
-            [
-                1.432950,
-                0.827314,
-                4.751000,
-                -1.432950,
-                0.827314,
-                4.751000,
-                0.0,
-                -1.654628,
-                4.751000,
-            ]
-        )
-        # ans = [[-1.432950, -2.481942, 0.0],
-        #       [-2.8659, 0.0, 0.0],
-        #       [-1.432950, -0.827314, -4.751000]]
-        ans = [
+        latt = Lattice([1.432950, 0.827314, 4.751000, -1.432950, 0.827314, 4.751000, 0.0, -1.654628, 4.751000])
+        expected = [
             [-1.43295, -2.481942, 0.0],
             [-2.8659, 0.0, 0.0],
             [-1.43295, -0.827314, -4.751],
         ]
-        self.assert_all_close(latt.get_niggli_reduced_lattice().matrix, ans)
+        self.assert_all_close(latt.get_niggli_reduced_lattice().matrix, expected)
 
         latt = Lattice.from_parameters(7.365450, 6.199506, 5.353878, 75.542191, 81.181757, 156.396627)
-        ans = [
+        expected = [
             [2.578932, 0.826965, 0.000000],
             [-0.831059, 2.067413, 1.547813],
             [-0.458407, -2.480895, 1.129126],
         ]
-        self.assert_all_close(latt.get_niggli_reduced_lattice().matrix, np.array(ans), 5)
+        self.assert_all_close(latt.get_niggli_reduced_lattice().matrix, np.array(expected), 5)
 
     def test_find_mapping(self):
         m = np.array([[0.1, 0.2, 0.3], [-0.1, 0.2, 0.7], [0.6, 0.9, 0.2]])
@@ -274,7 +244,7 @@ class LatticeTestCase(PymatgenTest):
 
         latt2 = Lattice(np.dot(rot, np.dot(scale, m).T).T)
         (aligned_out, rot_out, scale_out) = latt2.find_mapping(latt)
-        assert round(abs(abs(np.linalg.det(rot)) - 1), 7) == 0
+        assert abs(np.linalg.det(rot)) == approx(1)
 
         rotated = SymmOp.from_rotation_and_translation(rot_out).operate_multi(latt.matrix)
 
@@ -364,7 +334,7 @@ class LatticeTestCase(PymatgenTest):
             lattice.dot(np.zeros(6), np.zeros(8))
 
         # Passing vectors of different length is invalid.
-        with pytest.raises(ValueError):
+        with pytest.raises(ValueError, match="Coordinates must have same length"):
             lattice.dot(np.zeros(3), np.zeros(6))
 
     def test_get_points_in_sphere(self):
@@ -436,10 +406,8 @@ class LatticeTestCase(PymatgenTest):
         self.assert_all_close(output3, expected_pbc, 3)
 
     def test_monoclinic(self):
-        a, b, c, alpha, beta, gamma = self.monoclinic.parameters
-        assert round(abs(beta - 90), 7) != 0
-        assert round(abs(alpha - 90), 7) == 0
-        assert round(abs(gamma - 90), 7) == 0
+        assert self.monoclinic.angles == approx([90, 66, 90])
+        assert self.monoclinic.lengths == approx([10, 20, 30])
 
     def test_is_hexagonal(self):
         assert not self.cubic.is_hexagonal()

--- a/pymatgen/core/tests/test_operations.py
+++ b/pymatgen/core/tests/test_operations.py
@@ -41,7 +41,7 @@ class SymmOpTestCase(PymatgenTest):
         point = np.random.rand(3)
         new_coord = refl.operate(point)
         # Distance to the plane should be negatives of each other.
-        assert round(abs(np.dot(new_coord - origin, normal) - -np.dot(point - origin, normal)), 7) == 0
+        assert np.allclose(np.dot(new_coord - origin, normal), -np.dot(point - origin, normal))
 
     def test_apply_rotation_only(self):
         point = np.random.rand(3)

--- a/pymatgen/core/tests/test_periodic_table.py
+++ b/pymatgen/core/tests/test_periodic_table.py
@@ -165,7 +165,7 @@ class ElementTestCase(PymatgenTest):
         for k, v in testsets.items():
             assert Element(k).valence == v
 
-        with pytest.raises(ValueError):
+        with pytest.raises(ValueError, match="U has ambiguous valence"):
             Element("U").valence
 
         valence = Element("He").valence
@@ -276,8 +276,8 @@ class ElementTestCase(PymatgenTest):
         ]
 
         # Test all elements up to Uranium
-        for i in range(1, 104):
-            el = Element.from_Z(i)
+        for idx in range(1, 104):
+            el = Element.from_Z(idx)
             d = el.data
             for k in keys:
                 k_str = k.capitalize().replace("_", " ")
@@ -288,7 +288,7 @@ class ElementTestCase(PymatgenTest):
                 elif k == "iupac_ordering":
                     assert "IUPAC ordering" in d
                     assert getattr(el, k) is not None
-            el = Element.from_Z(i)
+            el = Element.from_Z(idx)
             if len(el.oxidation_states) > 0:
                 assert max(el.oxidation_states) == el.max_oxidation_state
                 assert min(el.oxidation_states) == el.min_oxidation_state
@@ -300,9 +300,9 @@ class ElementTestCase(PymatgenTest):
             Element.from_Z(1000)
 
     def test_ie_ea(self):
-        assert round(abs(Element.Fe.ionization_energies[2] - 30.651), 7) == 0
+        assert Element.Fe.ionization_energies[2] == pytest.approx(30.651)
         assert Element.Fe.ionization_energy == Element.Fe.ionization_energies[0]
-        assert round(abs(Element.Br.electron_affinity - 3.3635883), 7) == 0
+        assert Element.Br.electron_affinity == pytest.approx(3.3635883)
 
     def test_oxidation_states(self):
         el = Element.Fe
@@ -360,13 +360,13 @@ class SpeciesTestCase(PymatgenTest):
         self.specie4 = Species("Fe", 2, {"spin": 5})
 
     def test_init(self):
-        with pytest.raises(ValueError):
+        with pytest.raises(ValueError, match="magmom is not a supported property"):
             Species("Fe", 2, {"magmom": 5})
 
     def test_ionic_radius(self):
         assert self.specie2.ionic_radius == 78.5 / 100
         assert self.specie3.ionic_radius == 92 / 100
-        assert round(abs(Species("Mn", 4).ionic_radius - 0.67), 7) == 0
+        assert Species("Mn", 4).ionic_radius == pytest.approx(0.67)
 
     def test_eq(self):
         assert self.specie1 == self.specie3, "Static and actual constructor gives unequal result!"
@@ -413,15 +413,13 @@ class SpeciesTestCase(PymatgenTest):
         assert Species("Ni", 3).get_crystal_field_spin(spin_config="low") == 1
         assert Species("Ni", 4).get_crystal_field_spin(spin_config="low") == 0
 
-        with pytest.raises(AttributeError):
-            Species("Li", 1).get_crystal_field_spin()
-        with pytest.raises(AttributeError):
-            Species("Ge", 4).get_crystal_field_spin()
-        with pytest.raises(AttributeError):
-            Species("H", 1).get_crystal_field_spin()
-        with pytest.raises(AttributeError):
+        for elem in ("Li+", "Ge4+", "H+"):
+            symbol = Species(elem).symbol
+            with pytest.raises(AttributeError, match=f"Invalid element {symbol} for crystal field calculation"):
+                Species(elem).get_crystal_field_spin()
+        with pytest.raises(AttributeError, match="Invalid oxidation state 10 for element Fe"):
             Species("Fe", 10).get_crystal_field_spin()
-        with pytest.raises(ValueError):
+        with pytest.raises(ValueError, match="Invalid coordination or spin config"):
             Species("Fe", 2).get_crystal_field_spin("hex")
 
         s = Species("Co", 3).get_crystal_field_spin("tet", spin_config="low")

--- a/pymatgen/core/tests/test_sites.py
+++ b/pymatgen/core/tests/test_sites.py
@@ -4,6 +4,7 @@ import pickle
 
 import numpy as np
 import pytest
+from pytest import approx
 
 from pymatgen.core.composition import Composition
 from pymatgen.core.lattice import Lattice
@@ -26,6 +27,8 @@ class SiteTest(PymatgenTest):
         self.dummy_site = Site("X", [0, 0, 0])
 
     def test_properties(self):
+        assert not self.disordered_site.is_ordered
+        # TODO (janosh): this doesn't raise the expected error (match="specie property only works for ordered sites")
         with pytest.raises(AttributeError):
             self.disordered_site.specie
         assert isinstance(self.ordered_site.specie, Element)
@@ -58,8 +61,8 @@ class SiteTest(PymatgenTest):
 
     def test_distance(self):
         osite = self.ordered_site
-        assert round(abs(np.linalg.norm([0.25, 0.35, 0.45]) - osite.distance_from_point([0, 0, 0])), 7) == 0
-        assert round(abs(osite.distance(self.disordered_site) - 0), 7) == 0
+        assert np.linalg.norm([0.25, 0.35, 0.45]) == osite.distance_from_point([0, 0, 0])
+        assert osite.distance(self.disordered_site) == 0
 
     def test_pickle(self):
         o = pickle.dumps(self.propertied_site)
@@ -112,19 +115,18 @@ class PeriodicSiteTest(PymatgenTest):
 
     def test_distance(self):
         other_site = PeriodicSite("Fe", np.array([0, 0, 0]), self.lattice)
-        assert round(abs(self.site.distance(other_site) - 6.22494979899), 5) == 0
+        assert self.site.distance(other_site) == approx(6.22494979899)
 
     def test_distance_from_point(self):
-        assert round(abs(self.site.distance_from_point([0.1, 0.1, 0.1]) - 6.22494979899), 5) != 0
-        assert round(abs(self.site.distance_from_point([0.1, 0.1, 0.1]) - 6.0564015718906887), 5) == 0
+        assert self.site.distance_from_point([0.1, 0.1, 0.1]) == approx(6.0564015718906887)
 
     def test_distance_and_image(self):
         other_site = PeriodicSite("Fe", np.array([1, 1, 1]), self.lattice)
         distance, image = self.site.distance_and_image(other_site)
-        assert round(abs(distance - 6.22494979899), 5) == 0
+        assert distance == approx(6.22494979899)
         assert ([-1, -1, -1] == image).all()
         distance, image = self.site.distance_and_image(other_site, [1, 0, 0])
-        assert round(abs(distance - 19.461500456028563), 5) == 0
+        assert distance == approx(19.461500456028563)
         # Test that old and new distance algo give the same ans for
         # "standard lattices"
         lattice = Lattice([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
@@ -146,7 +148,7 @@ class PeriodicSiteTest(PymatgenTest):
         site = PeriodicSite("Fe", [0.1, 0.1, 0.1], latt)
         site2 = PeriodicSite("Fe", [0.99, 0.99, 0.99], latt)
         dist, img = site.distance_and_image(site2)
-        assert round(abs(dist - 0.15495358379511573), 7) == 0
+        assert dist == approx(0.15495358379511573)
         assert list(img) == [-11, 6, 0]
 
     def test_is_periodic_image(self):

--- a/pymatgen/core/tests/test_spectrum.py
+++ b/pymatgen/core/tests/test_spectrum.py
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 import numpy as np
+from pytest import approx
 from scipy import stats
 
 from pymatgen.core.spectrum import Spectrum
@@ -17,17 +18,17 @@ class SpectrumTest(PymatgenTest):
 
     def test_normalize(self):
         self.spec1.normalize(mode="max")
-        assert round(abs(np.max(self.spec1.y) - 1), 7) == 0
+        assert np.max(self.spec1.y) == approx(1)
         self.spec1.normalize(mode="sum")
-        assert round(abs(np.sum(self.spec1.y) - 1), 7) == 0
+        assert np.sum(self.spec1.y) == approx(1)
 
         self.multi_spec1.normalize(mode="sum")
-        assert round(abs(np.sum(self.multi_spec1.y[:, 0]) - 1), 7) == 0
-        assert round(abs(np.sum(self.multi_spec1.y[:, 1]) - 1), 7) == 0
+        assert np.sum(self.multi_spec1.y[:, 0]) == approx(1)
+        assert np.sum(self.multi_spec1.y[:, 1]) == approx(1)
 
         # XRD style mode.
         self.spec1.normalize(mode="max", value=100)
-        assert round(abs(np.max(self.spec1.y) - 100), 7) == 0
+        assert np.max(self.spec1.y) == approx(100)
 
     def test_operators(self):
         scaled_spect = 3 * self.spec1 + self.spec2

--- a/pymatgen/core/tests/test_units.py
+++ b/pymatgen/core/tests/test_units.py
@@ -2,6 +2,7 @@ from __future__ import annotations
 
 import pytest
 from numpy.testing import assert_array_equal
+from pytest import approx
 
 from pymatgen.core.units import (
     ArrayWithUnit,
@@ -41,42 +42,42 @@ class FloatWithUnitTest(PymatgenTest):
     def test_energy(self):
         a = Energy(1.1, "eV")
         b = a.to("Ha")
-        assert round(abs(b - 0.0404242579378), 7) == 0
+        assert b == approx(0.0404242579378)
         c = Energy(3.14, "J")
-        assert round(abs(c.to("eV") - 1.9598338493806797e19), 7) == 0
+        assert c.to("eV") == approx(1.9598338493806797e19)
         with pytest.raises(UnitError):
             Energy(1, "m")
 
         d = Energy(1, "Ha")
-        assert round(abs(a + d - 28.311386245987997), 7) == 0
-        assert round(abs(a - d - -26.111386245987994), 7) == 0
+        assert a == approx(d - 28.311386245987997)
+        assert a == approx(d - -26.111386245987994)
         assert a + 1 == 2.1
         assert str(a / d) == "1.1 eV Ha^-1"
 
         e = Energy(1, "kJ")
         f = e.to("kCal")
-        assert round(abs(f - 0.2390057361376673), 7) == 0
+        assert f == approx(0.2390057361376673)
         assert str(e + f) == "2.0 kJ"
         assert str(f + e) == "0.4780114722753346 kCal"
 
     def test_time(self):
         a = Time(20, "h")
-        assert round(abs(float(a.to("s")) - 3600 * 20), 7) == 0
+        assert float(a.to("s")) == approx(3600 * 20)
         # Test left and right multiplication.
         b = a * 3
-        assert round(abs(float(b) - 60.0), 7) == 0
+        assert float(b) == approx(60)
         assert str(b.unit) == "h"
         assert float(3 * a) == 60.0
         a = Time(0.5, "d")
-        assert round(abs(float(a.to("s")) - 3600 * 24 * 0.5), 7) == 0
+        assert float(a.to("s")) == approx(3600 * 24 * 0.5)
 
     def test_length(self):
         x = Length(4.2, "ang")
-        assert round(abs(x.to("cm") - 4.2e-08), 7) == 0
+        assert x.to("cm") == approx(4.2e-08)
         assert x.to("pm") == 420
         assert str(x / 2) == "2.1 ang"
         y = x**3
-        assert round(abs(y - 74.088), 7) == 0
+        assert y == approx(74.088)
         assert str(y.unit) == "ang^3"
 
     def test_memory(self):
@@ -135,17 +136,17 @@ class FloatWithUnitTest(PymatgenTest):
         e = Mass(1, "kg") * g * Length(1, "m")
         assert str(e) == "10.0 N m"
         form_e = FloatWithUnit(10, unit="kJ mol^-1").to("eV atom^-1")
-        assert round(abs(float(form_e) - 0.103642691905), 7) == 0
+        assert form_e == approx(0.103642691905)
         assert str(form_e.unit) == "eV atom^-1"
         with pytest.raises(UnitError):
             form_e.to("m s^-1")
         a = FloatWithUnit(1.0, "Ha^3")
         b = a.to("J^3")
-        assert round(abs(b - 8.28672661615e-53), 7) == 0
+        assert b == approx(8.28672661615e-53)
         assert str(b.unit) == "J^3"
         a = FloatWithUnit(1.0, "Ha bohr^-2")
         b = a.to("J m^-2")
-        assert round(abs(b - 1556.8931028218924), 7) == 0
+        assert b == approx(1556.8931028218924)
         assert str(b.unit) == "J m^-2"
 
     def test_as_base_units(self):
@@ -168,14 +169,14 @@ class ArrayWithFloatWithUnitTest(PymatgenTest):
         """
         a = EnergyArray(1.1, "eV")
         b = a.to("Ha")
-        assert round(abs(float(b) - 0.0404242579378), 7) == 0
+        assert (b) == approx(0.0404242579378)
         c = EnergyArray(3.14, "J")
-        assert round(abs(float(c.to("eV")) - 1.9598338493806797e19), 5) == 0
+        assert (c.to("eV")) == approx(1.9598338493806797e19)
         # self.assertRaises(ValueError, Energy, 1, "m")
 
         d = EnergyArray(1, "Ha")
-        assert round(abs(float(a + d) - 28.311386245987997), 7) == 0
-        assert round(abs(float(a - d) - -26.111386245987994), 7) == 0
+        assert (a + d) == approx(28.311386245987997)
+        assert (a - d) == approx(-26.111386245987994)
         assert float(a + 1) == 2.1
 
     def test_time(self):
@@ -183,8 +184,7 @@ class ArrayWithFloatWithUnitTest(PymatgenTest):
         Similar to FloatWithUnitTest.test_time.
         Check whether EnergyArray and FloatWithUnit have same behavior.
         """
-        # here there's a minor difference because we have a ndarray with
-        # dtype=int.
+        # here there's a minor difference because we have a ndarray with dtype=int
         a = TimeArray(20, "h")
         assert a.to("s") == 3600 * 20
         # Test left and right multiplication.
@@ -197,7 +197,7 @@ class ArrayWithFloatWithUnitTest(PymatgenTest):
         Check whether EnergyArray and FloatWithUnit have same behavior.
         """
         x = LengthArray(4.2, "ang")
-        assert round(abs(float(x.to("cm")) - 4.2e-08), 7) == 0
+        assert float(x.to("cm")) == approx(4.2e-08)
         assert float(x.to("pm")) == 420
         assert str(x / 2) == "2.1 ang"
 

--- a/pymatgen/core/tests/test_units.py
+++ b/pymatgen/core/tests/test_units.py
@@ -49,8 +49,8 @@ class FloatWithUnitTest(PymatgenTest):
             Energy(1, "m")
 
         d = Energy(1, "Ha")
-        assert a == approx(d - 28.311386245987997)
-        assert a == approx(d - -26.111386245987994)
+        assert a + d == approx(28.311386245987997)
+        assert a - d == approx(-26.111386245987994)
         assert a + 1 == 2.1
         assert str(a / d) == "1.1 eV Ha^-1"
 

--- a/pymatgen/io/gaussian.py
+++ b/pymatgen/io/gaussian.py
@@ -371,45 +371,45 @@ class GaussianInput:
         with zopen(filename, "r") as f:
             return GaussianInput.from_string(f.read())
 
-    def _find_nn_pos_before_site(self, siteindex):
+    def _find_nn_pos_before_site(self, site_idx):
         """
         Returns index of nearest neighbor atoms.
         """
-        alldist = [(self._mol.get_distance(siteindex, i), i) for i in range(siteindex)]
-        alldist = sorted(alldist, key=lambda x: x[0])
-        return [d[1] for d in alldist]
+        all_dist = [(self._mol.get_distance(site_idx, idx), idx) for idx in range(site_idx)]
+        all_dist = sorted(all_dist, key=lambda x: x[0])
+        return [d[1] for d in all_dist]
 
     def get_zmatrix(self):
         """
         Returns a z-matrix representation of the molecule.
         """
         output = []
-        outputvar = []
-        for i, site in enumerate(self._mol):
-            if i == 0:
+        output_var = []
+        for idx, site in enumerate(self._mol):
+            if idx == 0:
                 output.append(f"{site.specie}")
-            elif i == 1:
-                nn = self._find_nn_pos_before_site(i)
-                bondlength = self._mol.get_distance(i, nn[0])
-                output.append(f"{self._mol[i].specie} {nn[0] + 1} B{i}")
-                outputvar.append(f"B{i}={bondlength:.6f}")
-            elif i == 2:
-                nn = self._find_nn_pos_before_site(i)
-                bondlength = self._mol.get_distance(i, nn[0])
-                angle = self._mol.get_angle(i, nn[0], nn[1])
-                output.append(f"{self._mol[i].specie} {nn[0] + 1} B{i} {nn[1] + 1} A{i}")
-                outputvar.append(f"B{i}={bondlength:.6f}")
-                outputvar.append(f"A{i}={angle:.6f}")
+            elif idx == 1:
+                nn = self._find_nn_pos_before_site(idx)
+                bond_length = self._mol.get_distance(idx, nn[0])
+                output.append(f"{self._mol[idx].specie} {nn[0] + 1} B{idx}")
+                output_var.append(f"B{idx}={bond_length:.6f}")
+            elif idx == 2:
+                nn = self._find_nn_pos_before_site(idx)
+                bond_length = self._mol.get_distance(idx, nn[0])
+                angle = self._mol.get_angle(idx, nn[0], nn[1])
+                output.append(f"{self._mol[idx].specie} {nn[0] + 1} B{idx} {nn[1] + 1} A{idx}")
+                output_var.append(f"B{idx}={bond_length:.6f}")
+                output_var.append(f"A{idx}={angle:.6f}")
             else:
-                nn = self._find_nn_pos_before_site(i)
-                bondlength = self._mol.get_distance(i, nn[0])
-                angle = self._mol.get_angle(i, nn[0], nn[1])
-                dih = self._mol.get_dihedral(i, nn[0], nn[1], nn[2])
-                output.append(f"{self._mol[i].specie} {nn[0] + 1} B{i} {nn[1] + 1} A{i} {nn[2] + 1} D{i}")
-                outputvar.append(f"B{i}={bondlength:.6f}")
-                outputvar.append(f"A{i}={angle:.6f}")
-                outputvar.append(f"D{i}={dih:.6f}")
-        return "\n".join(output) + "\n\n" + "\n".join(outputvar)
+                nn = self._find_nn_pos_before_site(idx)
+                bond_length = self._mol.get_distance(idx, nn[0])
+                angle = self._mol.get_angle(idx, nn[0], nn[1])
+                dih = self._mol.get_dihedral(idx, nn[0], nn[1], nn[2])
+                output.append(f"{self._mol[idx].specie} {nn[0] + 1} B{idx} {nn[1] + 1} A{idx} {nn[2] + 1} D{idx}")
+                output_var.append(f"B{idx}={bond_length:.6f}")
+                output_var.append(f"A{idx}={angle:.6f}")
+                output_var.append(f"D{idx}={dih:.6f}")
+        return "\n".join(output) + "\n\n" + "\n".join(output_var)
 
     def get_cart_coords(self) -> str:
         """Return the Cartesian coordinates of the molecule"""

--- a/pymatgen/util/coord.py
+++ b/pymatgen/util/coord.py
@@ -304,9 +304,9 @@ def lattice_points_in_supercell(supercell_matrix):
 
     frac_points = np.dot(all_points, np.linalg.inv(supercell_matrix))
 
-    tvects = frac_points[np.all(frac_points < 1 - 1e-10, axis=1) & np.all(frac_points >= -1e-10, axis=1)]
-    assert len(tvects) == round(abs(np.linalg.det(supercell_matrix)))
-    return tvects
+    t_vecs = frac_points[np.all(frac_points < 1 - 1e-10, axis=1) & np.all(frac_points >= -1e-10, axis=1)]
+    assert len(t_vecs) == round(abs(np.linalg.det(supercell_matrix)))
+    return t_vecs
 
 
 def barycentric_coords(coords, simplex):

